File: src/main/scala/nl/rug/jbi/search/ssap/util/JarContainer.scala
package nl.rug.jbi.search.ssap.util

import java.io.{InputStream, File}
import java.util.jar.JarFile
import scala.collection.convert.wrapAsScala.enumerationAsScalaIterator

/**
  * Container for .class files in a .jar.
  *
  * @author Daniel Feitosa
  */
class JarContainer(project: File) extends ScalaProjectContainer(project) {
  private val jar = new JarFile(project)

  override def isValid = JarContainer.isValid(project)
  override def forEachClass[A](callback: (InputStream) => A) =
    jar.entries().foreach( e => callback(getClassStream(e.getName)) )
  override def getClassStream(classname: String) = jar.getInputStream(jar.getJarEntry(classname))
  override def close = jar.close
}

object JarContainer {
  final def isValid(project: File) = project.isFile && project.getName.matches("(?i).*\\.jar$")
}

File: ssap/src/main/java/nl/rug/jbi/search/ssap/util/JarContainer.java
package nl.rug.jbi.search.ssap.util;

import java.io.*;
import java.util.Iterator;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;

/**
 * Container for .class files in a .jar.
 *
 * @author Andrei Dumitriu,  Daniel Feitosa
 */
public class JarContainer extends ProjectContainer{
    private final JarFile jar;
    public JarContainer(File project) throws IOException {
        super(project);
        jar = new JarFile(project);
    }

    @Override
    public Boolean isValid() {
        return isValid(project);
    }

    private static Boolean isValid(File project) {
        return project.isFile() && project.getName().matches("(?i).*\\.jar$");
    }

    @Override
    public <A> void forEachClass(CallbackFunction<A> callback) throws IOException {
        for (Iterator<JarEntry> it = jar.entries().asIterator(); it.hasNext(); ) {
            JarEntry entry = it.next();
            callback.run(getClassStream(entry.getName()));
        }
    }

    @Override
    public InputStream getClassStream(String classname) throws IOException {
        return jar.getInputStream(jar.getJarEntry(classname));
    }

    @Override
    public void close() throws IOException {
        jar.close();
    }
}
